### PR TITLE
Resolves #1391: Implement TypedArray updates to index handling.

### LINKED ISSUE #1391: Implement TypedArray updates to index handling
[ECMAScript unified the handling](https://github.com/tc39/ecma262/pull/410) of indexes across all the TypedArray and related data structures. A number of minor semantics have changed.

### PR DESCRIPTION
**Pull request by atulkatti**
Resolves #1391: Implement TypedArray updates to index handling.
1. Also, made changes to ArrayBuffer and DataView as per the spec updates.
2. Added some more tests for DataView.

### CODE REVIEW

**Comment by atulkatti** [comment]
@Microsoft/chakra-developers

File: test/typedarray/dataview.js [hunk 1]
```diff
@@ -104,7 +103,7 @@ function testOneOffset(dataView, offSet, value)
                 print(getFuncs[j] + " = " +  result);
                 }
 
-            print("set little endian value offset " + offSet + " value " + value + " method " + setFuncs[i]);
+            print("set big endian value offset " + offSet + " value " + value + " method " + setFuncs[i]);
```
Comment (atulkatti) [inline test/typedarray/dataview.js]: You can ignore the changes to dataview.js, dataview.baseline and samethread.baseline. I fixed this typo in the comments of this test and that caused a lot of changes to the baseline. No other changes to these files.

**Comment by tcare** [comment]
Make sure to prefix the issue number with # so it triggers GitHub to link them.

File: lib/Runtime/Library/TypedArray.cpp [hunk 2]
```diff
@@ -509,12 +510,15 @@ namespace Js
             if (args.Info.Count > 3 && !JavascriptOperators::IsUndefinedObject(args[3]))
             {
                 mappedLength = ArrayBuffer::ToIndex(args[3], JSERR_InvalidTypedArrayLength, scriptContext, ArrayBuffer::MaxArrayBufferLength / elementSize, false);
+                newByteLength = mappedLength * elementSize;
 
-                if ((uint32)mappedLength > (byteLength - offset)/ elementSize)
+                if (offset + newByteLength > byteLength)
```
Comment (tcare) [inline lib/Runtime/Library/TypedArray.cpp]: Which case was this fixing?

Comment (atulkatti) [inline lib/Runtime/Library/TypedArray.cpp]: I noticed a slight change in the spec while reviewing the indexing related changes. Please refer to the spec at https://tc39.github.io/ecma262/#sec-typedarray-buffer-byteoffset-length section 22.2.4.5 step 12.

**Comment by tcare** [comment]
@boingoing please take a look

**Comment by tcare** [comment]
@atulkatti please run buddy testing as per the email I sent

**Comment by ljharb** [comment]
(Reverted by #2350)

**Comment by dilijev** [comment]
@ljharb @tcare FYI https://github.com/Microsoft/ChakraCore/issues/1391#issuecomment-272012172

**Pull request merged**